samza-core/src/main/java/org/apache/samza/coordinator/LeaderElector.java

```diff
@@ -32,11 +32,16 @@
 @InterfaceStability.Evolving
 public interface LeaderElector {
   /**
-   * Async method that helps the caller participate in leader election.
+   * Register a LeaderElectorListener
    *
-   * @param leaderElectorListener to be invoked if the caller is chosen as a leader through the leader election process
+   * @param listener {@link LeaderElectorListener} interfaces to be invoked upon completion of leader election participation
    */
-  void tryBecomeLeader(LeaderElectorListener leaderElectorListener);
+  void setLeaderElectorListener(LeaderElectorListener listener);
```

Comment:
1. It just makes the code more modularized and testable. 
2. Aside from that, if you are passing some kind of state within a recursive call, you need to manage the lifecycle of that instance in recursive function. It makes the code more vulnerable to bugs. 
3. Speaking of semantics, it is the same callback function that needs to be invoked in every call. So, there is really no need to pass around a listener that doesn't do anything different. Does this make sense?